### PR TITLE
Add support for SUSE OVAL in Vulnerability Detector

### PR DESCRIPTION
**Pull request by MarcelKemp**
|Related issue|
|---|
|#7171|

## Description

This pull request adds the following features:

- We have added support for the following OSs:

  - [SLES11](https://ftp.suse.com/pub/projects/security/oval/suse.linux.enterprise.server.11.xml), [SLES12](https://ftp.suse.com/pub/projects/security/oval/suse.linux.enterprise.server.12.xml) and [SLES15](https://ftp.suse.com/pub/projects/security/oval/suse.linux.enterprise.server.15.xml)
  - [SLED11](https://ftp.suse.com/pub/projects/security/oval/suse.linux.enterprise.desktop.11.xml), [SLED12](https://ftp.suse.com/pub/projects/security/oval/suse.linux.enterprise.desktop.12.xml) and [SLED15](https://ftp.suse.com/pub/projects/security/oval/suse.linux.enterprise.desktop.15.xml)

- Functions have been created to resolve dependencies linked to vulnerable packages. Being parsed the _xml_ feeds and inserted in tables both the vulnerabilities, as their relation with said dependencies.

- Vulnerability Detector now scans the programs collected by syscollector against vulnerabilities obtained from the NVD and SUSE OVAL. In SUSE's OVAL, as it has dependencies associated with vulnerabilities, first a `pre-scan` must be performed to find out the dependencies installed in the system, and then, the `scan` must be carried out taking into account said association.

## Configuration options

The Vulnerability Detector configuration has been extended to support the added provider:

```xml
<vulnerability-detector>
    ...
        <!-- SUSE OS vulnerabilities -->
    <provider name="suse">
      <enabled>no</enabled>
      <os>15-server</os>
      <os>15-desktop</os>
      <os>12-server</os>
      <os>12-desktop</os>
      <os>11-server</os>
      <os>11-desktop</os>
      <update_interval>1h</update_interval>
    </provider>
    ...
</vulnerability-detector>
````

## Logs/Alerts example

Here we can see the logs when fetching the SUSE OVAL feed:

```
2021/08/31 09:49:41 wazuh-modulesd:vulnerability-detector[21110] wm_vuln_detector.c:2855 at wm_vuldet_insert(): DEBUG: (5489): Inserting 'SUSE' vulnerabilities dependencies.
...
2021/08/31 09:50:05 wazuh-modulesd:vulnerability-detector[21110] wm_vuln_detector.c:4255 at wm_vuldet_index_feed(): DEBUG: (5427): Refresh of 'SUSE Linux Enterprise Server 15' database finished.
2021/08/31 09:50:05 wazuh-modulesd:vulnerability-detector[21110] wm_vuln_detector.c:4606 at wm_vuldet_check_feed(): INFO: (5430): The update of the 'SUSE Linux Enterprise Server 15' feed finished successfully.
```

Vulnerability log displayed during Amazon Linux packages pre-scan and scan:

```
2021/08/31 12:16:56 wazuh-modulesd:vulnerability-detector[36160] wm_vuln_detector.c:2277 at wm_vuldet_linux_oval_vulnerabilities(): DEBUG: (5458): Package 'curl' inserted into the vulnerability 'CVE-2020-8177'. Version (7.60.0-3.17.1) 'less than' '7.60.0-3.29.1' (feed 'OVAL').
2021/08/31 12:17:24 wazuh-modulesd:vulnerability-detector[36160] wm_vuln_detector.c:1619 at wm_vuldet_send_cve_report(): DEBUG: (5468): The 'curl' package (7.60.0-3.17.1) from agent '014' is vulnerable to 'CVE-2020-8177'. Condition: 'Package less than 7.60.0-3.29.1'
```

The alert generated by the above vulnerability:

```json
{
...
"description":"CVE-2020-8177 affects curl",
"gdpr":["IV_35.7.d"],
"pci_dss":["11.2.1","11.2.3"],
"tsc":["CC7.1","CC7.2"]},
"agent":{"id":"014","name":"suse-enterprise15","ip":"10.0.2.15"},
...
"vulnerability":{
"package":
{"name":"curl",
"version":"7.60.0-3.17.1",
"architecture":"x86_64",
"condition":"Package less than 7.60.0-3.29.1"},
"cvss":{
"cvss2":{
"vector":{
"attack_vector":"local",
"access_complexity":"low",
"authentication":"none",
"confidentiality_impact":"partial",
"integrity_impact":"partial",
"availability":"partial"},
"base_score":"4.600000"},
"cvss3":{
"vector":{
"attack_vector":"local",
"access_complexity":"low",
"privileges_required":"low",
"user_interaction":"none",
"scope":"unchanged",
"confidentiality_impact":"none",
"integrity_impact":"high",
"availability":"high"},
"base_score":"7.100000"}},
"cve":"CVE-2020-8177",
"title":"CVE-2020-8177 affects curl","rationale":"curl 7.20.0 through 7.70.0 is vulnerable to improper restriction of names for files and other resources that can lead too overwriting a local file when the -J flag is used.",
"severity":"High",
"published":"2020-12-14",
"updated":"2021-04-01",
"cwe_reference":"CWE-74",
"references":
...
,"assigner":"cve-assignments@hackerone.com",
"cve_version":"4.0"}},
"location":"vulnerability-detector"}
```

And how to fix the package vulnerability (in the vulnerable agent):

```
root@suse-enterprise15 vagrant $ zypper update curl
Refreshing service 'Basesystem_Module_15_SP1_x86_64'.
Refreshing service 'Desktop_Applications_Module_15_SP1_x86_64'.
Refreshing service 'SUSE_Linux_Enterprise_Server_15_SP1_x86_64'.
Refreshing service 'Server_Applications_Module_15_SP1_x86_64'.
Loading repository data...
Reading installed packages...
Resolving package dependencies...

The following package is going to be upgraded:
  curl

1 package to upgrade.
Overall download size: 296.2 KiB. Already cached: 0 B. After the operation, 114.0 B will be freed.
Continue? [y/n/v/...? shows all options] (y): y
Retrieving package curl-7.60.0-3.35.1.x86_64                    (1/1), 296.2 KiB (491.1 KiB unpacked)
Retrieving: curl-7.60.0-3.35.1.x86_64.rpm .....................................................[done]
Checking for file conflicts: ..................................................................[done]
(1/1) Installing: curl-7.60.0-3.35.1.x86_64 ...................................................[done]
```

## Tests

- Compilation without warnings in every supported platform
  - [x] Linux
  - [ ] Windows
  - [ ] MAC OS X
- [x] Source installation
- [ ] Package installation
- [x] Source upgrade
- [ ] Package upgrade
- [x] Review logs syntax and correct language
- [ ] QA templates contemplate the added capabilities

- Memory tests for Linux
  - [x] Scan-build report
  - [x] Coverity
  - [x] Valgrind (memcheck and descriptor leaks check)
  - [ ] Dr. Memory
  - [ ] AddressSanitizer

- [x] Retrocompatibility with older Wazuh versions
- [ ] Working on cluster environments
- [x] Configuration on demand reports new parameters
- [x] The data flow works as expected (agent-manager-api-app)
- [x] Added unit tests (for new features)
- [ ] Stress test for affected components

### CODE REVIEW

**Comment by StefanSa** [comment]
Hi there,
only SLES, or also openLeap (openuse) ?
It would be good, if it would also apply to openLeap.

**Comment by chemamartinez** [comment]
Hi @StefanSa,

Our goal was to add support to Suse Linux Enterprise and OpenSUSE. Unfortunately, the available [OVAL feeds](https://www.suse.com/support/security/oval/) only apply to the first one, so evaluating these vulnerabilities against OpenSUSE packages would lead to unexpected results.

For now, it is not possible to add native support to OpenSUSE until we find OVAL feeds for these platforms.

Regards.

File: src/unit_tests/wazuh_modules/vulnerability_detector/test_wm_vuln_detector.c [hunk 1]
```diff
@@ -838,6 +842,144 @@ static int teardown_info_cves(void **state) {
...
 }
 
+static int setup_dependencies(void **state) {
+    dependencies * deps_it = NULL;
+    os_calloc(1, sizeof(dependencies), deps_it);
+    *state = deps_it;
+
+    deps_it->elements = 3;
+    os_calloc(3, sizeof(char *), deps_it->test_ref);
+    os_strdup("dependencie:id:1", deps_it->test_ref[0]);
```
Comment (chemamartinez) [inline src/unit_tests/wazuh_modules/vulnerability_detector/test_wm_vuln_detector.c:852]: Please change `dependencie` by `dependency` or `dependencies`

**Review by chemamartinez — CHANGES_REQUESTED** [review]

**Comment by chemamartinez** [comment]
While reviewing the reports generated for SUSE agents I noticed a condition that leads to false negatives.

For SUSE agents, it is expected to have several entries in the vulnerability database for the same CVE and package, since they apply in different situations according to the installed dependencies.

Here is an example with the *CVE-2019-16234*:

**Comment by chemamartinez** [comment]
Fixed

```
2021/09/30 16:34:29 wazuh-modulesd:vulnerability-detector[172503] wm_vuln_detector.c:2286 at wm_vuldet_linux_oval_vulnerabilities(): DEBUG: (5458): Package 'kernel-macros' inserted into the vulnerability 'CVE-2019-16234'. Version (4.12.14-197.7.1) 'less than' '4.12.14-197.26.1' (feed 'OVAL').
```

**Review by chemamartinez — APPROVED** [review]
LGTM!

**Review by chemamartinez — DISMISSED** [review]

**Review by chemamartinez — APPROVED** [review]

**Pull request merged**